src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+  auto specifierUrl = KJ_ASSERT_NONNULL(BundleBuilder::BASE.tryResolve(specifier));
+  Builder::alias(aliasUrl, specifierUrl);
+  return *this;
+}
+
+// ======================================================================================
+
+ModuleBundle::BuiltinBuilder::BuiltinBuilder(Type type)
+    : ModuleBundle::Builder(toModuleBuilderType(type)) {}
+
+ModuleBundle::BuiltinBuilder& ModuleBundle::BuiltinBuilder::addSynthetic(
+    const Url& specifier,
+    ModuleBundle::BundleBuilder::EvaluateCallback callback) {
+  ensureIsNotBundleSpecifier(specifier);
+  Builder::add(specifier,
+      [url = specifier.clone(),
+       callback = kj::mv(callback),
+       type = type()]
+      (const ResolveContext& context) mutable -> kj::Maybe<kj::Own<Module>> {
+    return Module::newSynthetic(kj::mv(url), type, kj::mv(callback));
+  });
+  return *this;
+}
+
+ModuleBundle::BuiltinBuilder& ModuleBundle::BuiltinBuilder::addEsm(
+    const Url& specifier,
+    kj::ArrayPtr<const char> source) {
+  ensureIsNotBundleSpecifier(specifier);
+  Builder::add(specifier, [url = specifier.clone(), source, type = type()]
+      (const ResolveContext& context) mutable -> kj::Maybe<kj::Own<Module>> {
+    return Module::newEsm(kj::mv(url), type, source);
+  });
+  return *this;
+}
+
+// ======================================================================================
+ModuleRegistry::Builder::Builder(const ResolveObserver& observer, Options options)
+     : observer(observer), options(options) {}
+
+bool ModuleRegistry::Builder::allowsFallback() const {
+  return (options & Options::ALLOW_FALLBACK) == Options::ALLOW_FALLBACK;
+}
+
+ModuleRegistry::Builder& ModuleRegistry::Builder::setParent(ModuleRegistry& parent) {
+  maybeParent = parent;
+  return *this;
+}
+
+ModuleRegistry::Builder& ModuleRegistry::Builder::add(kj::Own<ModuleBundle> bundle) {
+  if (!allowsFallback()) {
+    KJ_REQUIRE(bundle->type() != ModuleBundle::Type::FALLBACK,
+               "Fallback bundle types are not allowed for this registry");
+  }
+  bundles_[static_cast<int>(bundle->type())].add(kj::mv(bundle));
+  return *this;
+}
+
+kj::Own<ModuleRegistry> ModuleRegistry::Builder::finish() {
+  return kj::heap<ModuleRegistry>(this);
+}
+
+ModuleRegistry::ModuleRegistry(ModuleRegistry::Builder* builder)
+    : observer(builder->observer),
+      maybeParent(builder->maybeParent) {
+  bundles_[kBundle] = builder->bundles_[kBundle].releaseAsArray();
+  bundles_[kBuiltin] = builder->bundles_[kBuiltin].releaseAsArray();
+  bundles_[kBuiltinOnly] = builder->bundles_[kBuiltinOnly].releaseAsArray();
+  bundles_[kFallback] = builder->bundles_[kFallback].releaseAsArray();
+}
+
+void ModuleRegistry::attachToIsolate(
+    Lock& js,
+    const CompilationObserver& observer) {
+  // We use new here because it is required for the v8 integration.
+  // Responsibility for cleaning up the type is passed to v8 when
+  // the context is cleaned up.
+  new IsolateModuleRegistry(js, *this, observer);
```

Comment:
The unwrapped `new` here feels scary, but we keep track of this through the `SetAlignedPointerInEmbedderData` call in the `IsolateModuleRegistry` constructor right?

Reply:
Yes, this is a v8 requirement. V8 takes ownership of the pointer and destroys it when necessary.